packages/auth/src/grant/service.ts

```diff
@@ -0,0 +1,131 @@
+import { v4 } from 'uuid'
+import * as crypto from 'crypto'
+import { Transaction, TransactionOrKnex } from 'objection'
+
+import { BaseService } from '../shared/baseService'
+import { Grant, GrantState, StartMethod, FinishMethod } from './model'
+import { AccessRequest, isAccessRequest } from '../access/types'
+import { ClientInfo } from '../client/service'
+import { IAppConfig } from '../config/app'
+import { AccessService } from '../access/service'
+
+export interface GrantService {
+  validateGrantRequest(grantRequest: GrantRequest): boolean
+  initiateGrant(grantRequest: GrantRequest): Promise<GrantResponse>
+}
+
+interface ServiceDependencies extends BaseService {
+  accessService: AccessService
+  config: IAppConfig
+  knex: TransactionOrKnex
+}
+
+// TODO: maybe update docs.openpayments.guide with location
+export interface GrantRequest {
+  access: AccessRequest[]
+  client: ClientInfo
+  interact: {
+    start: StartMethod[]
+    finish: {
+      method: FinishMethod
+      uri: string
+      nonce: string
+    }
+  }
+}
+
+export interface GrantResponse {
+  interact: {
+    redirect: string
+    finish: string
+  }
+  continue: {
+    access_token: {
+      value: string
+    }
+    uri: string
+  }
+}
+
+export async function createGrantService({
+  logger,
+  accessService,
+  config,
+  knex
+}: ServiceDependencies): Promise<GrantService> {
+  const log = logger.child({
+    service: 'GrantService'
+  })
+  const deps: ServiceDependencies = {
+    logger: log,
+    accessService,
+    config,
+    knex
+  }
+  return {
+    validateGrantRequest: (grantRequest: GrantRequest) =>
+      validateGrantRequest(grantRequest),
+    initiateGrant: (grantRequest: GrantRequest, trx?: Transaction) =>
+      initiateGrant(deps, grantRequest, trx)
+  }
+}
+
+function validateGrantRequest(
+  grantRequest: GrantRequest
+): grantRequest is GrantRequest {
+  if (typeof grantRequest.access !== 'object') return false
+  for (const access of grantRequest.access) {
+    if (!isAccessRequest(access)) return false
+  }
+
+  return grantRequest.interact?.start !== undefined
+}
+
+async function initiateGrant(
+  deps: ServiceDependencies,
+  grantRequest: GrantRequest,
+  trx?: Transaction
+): Promise<GrantResponse> {
+  const { accessService, knex, config } = deps
+
+  const {
+    access,
+    interact: {
+      start,
+      finish: { method: finishMethod, uri: finishUri, nonce: clientNonce }
+    }
+  } = grantRequest
+
+  const grant = await Grant.query(trx || knex).insert({
```

Comment:
This should create a transaction if `trx` is undefined, so that the grant insert can be rolled back if any of the access inserts fail.
See: https://github.com/interledger/rafiki/blob/main/packages/backend/src/open_payments/payment/incoming/service.ts#L120-L145